### PR TITLE
cmd, eth: orchestrator activated/deactivated watcher

### LINKED ISSUE #1150: Add a TranscoderWatcher event watcher
**Is your feature request related to a problem? Please describe.**

Create a `TranscoderWatcher` in `eth/watchers` that watches for emitted `TranscoderActivated` and `TranscoderDeactivated` events in order to be able to determine the current active orchestrator set. 

The TranscoderWatcher will update the `ActivationRound`/`DeactivationRound` fields for the subject orchestrator in the database. 

**Describe the solution you'd like**
`TranscoderWatcher` watches for on-chain events emitted by the `BondingManager` pertaining to orchestrator activation and deactivation

- watch for `TranscoderActived` events
- watch for `TranscoderDeactivated` events

```
type transcoderWatcher struct {
    db      *common.DB
    dec     *EventDecoder
    lpEth   eth.LivepeerEthClient
    quit    chan struct{}
}
```

#### Handling events

- Event added: update values in DB with the `ActivationRound` and `DeactivationRound` arguments emitted by the event
- Event removed: Make an RPC call to `BondingManager.GetTranscoder()` and update the DB with the returned `ActivationRound` and `DeactivationRound` values

##### Watch

Start watching the `blockwatch` subscription for `TranscoderActivated` and `TranscoderDeactivated` events

```
func (tw *transcoderWatcher) Watch() error
```

##### Stop

Unsubscribe from `TranscoderActivated` and `TranscoderDeactivated` events propagated by `blockwatch` 

```
func (tw *transcoderWatcher) Stop()
```

**Additional context**

Full spec for active orchestrator set tracking: https://hackmd.io/@AO8Ceb4zScyJGk3uD6Tnhw/B1B-rsO9H

### PR DESCRIPTION
**Pull request by kyriediculous**
**What does this pull request do? Explain your changes. (required)**
This PR adds an implementation for an `OrchestratorWatcher` that watches for on-chain emitted `TranscoderActivated` and `TranscoderDeactivated` events and updates orchestrators stored in the database accordingly. 

**Specific updates (required)**
- Added `TranscoderActivated` and `TranscoderDeactivated` to `eth/watchers/topics.go`
- Added an `OrchestratorStore` interface to `eth/watchers/types` to be able to stub out `common.DB` for easier unit testing
- Added an `OrchestratorWatcher` type
- Start an `OrchestratorWatcher`  in `livepeer.go`
- Add a `StubOrchestratorStore` and helpers for creating stubbed out `TranscoderActivated/TranscoderDeactivated` events to `eth/watchers/stub.go`
- Added `ActivationRound` and `DeactivationRound` fields to `types.Transcoder`
- Make sure `LivepeerEthClient.GetTranscoder()` now also returns `ActivationRound` and `DeactivationRound`

**How did you test each of these updates (required)**
Ran unit tests

**Does this pull request close any open issues?**
Fixes #1150 

**Checklist:**
- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

**Comment by yondonfu** [comment]
Changed the target branch to `nv/db-orchestrator-updates` since this PR builds on top of that branch to make it easier to review the commits specific to this PR.

File: eth/watchers/orchestratorwatcher.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+func (ow *OrchestratorWatcher) handleLog(log types.Log) error {
+	eventName, err := ow.dec.FindEventName(log)
+	if err != nil {
+		// Noop if we cannot find the event name
+		return nil
+	}
+
+	switch eventName {
+	case "TranscoderActivated":
+		ow.handleTranscoderActivated(log)
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Let's handle the returned error

File: eth/watchers/orchestratorwatcher.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	case "TranscoderDeactivated":
+		ow.handleTranscoderDeactivated(log)
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Let's handle the returned error

File: eth/watchers/types.go [hunk 3]
```diff
@@ -13,3 +14,7 @@ type BlockWatcher interface {
 type EventWatcher interface {
 	Subscribe(sink chan<- types.Log) event.Subscription
 }
+
+type OrchestratorStore interface {
```
Comment (yondonfu) [inline eth/watchers/types.go]: Doesn't seem like we need to export this interface - can we make it internal to the package?

File: eth/watchers/orchestratorwatcher.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/common"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/eth/blockwatch"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+)
+
+type OrchestratorWatcher struct {
+	db      OrchestratorStore
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Could we name this something like `store` since the name `db` feels strongly tied to a database and the interface is not necessarily implemented by a database?

File: eth/watchers/stub.go [hunk 5]
```diff
@@ -150,54 +155,86 @@ func newStubWithdrawalLog() types.Log {
...
+	log.Address = stubBondingManagerAddr
+	transcoder := ethcommon.LeftPadBytes(stubTranscoder.Bytes(), 32)
+	var transcoderTopic ethcommon.Hash
+	copy(transcoderTopic[:], transcoder[:])
+	log.Topics = []ethcommon.Hash{
+		crypto.Keccak256Hash([]byte("TranscoderActivated(address,uint256)")),
+		transcoderTopic,
+	}
+	var data []byte
+	data = append(data, ethcommon.LeftPadBytes(stubActivationRound.Bytes(), 32)...)
```
Comment (yondonfu) [inline eth/watchers/stub.go]: Can shorten this to just `log.Data = ethcommon.LeftPadBytes(..)`

File: eth/watchers/stub.go [hunk 6]
```diff
@@ -150,54 +155,86 @@ func newStubWithdrawalLog() types.Log {
...
+	log.Address = stubBondingManagerAddr
+	transcoder := ethcommon.LeftPadBytes(stubTranscoder.Bytes(), 32)
+	var transcoderTopic ethcommon.Hash
+	copy(transcoderTopic[:], transcoder[:])
+	log.Topics = []ethcommon.Hash{
+		crypto.Keccak256Hash([]byte("TranscoderDeactivated(address,uint256)")),
+		transcoderTopic,
+	}
+	var data []byte
+	data = append(data, ethcommon.LeftPadBytes(stubDeactivationRound.Bytes(), 32)...)
```
Comment (yondonfu) [inline eth/watchers/stub.go]: Can shorten this to just `log.Data = ethcommon.LeftPadBytes(...)`

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: eth/watchers/stub.go (hunk 6 above)
Comment (kyriediculous) [reply eth/watchers/stub.go]: fixed in 6a397072

File: eth/watchers/stub.go (hunk 5 above)
Comment (kyriediculous) [reply eth/watchers/stub.go]: fixed in 6a397072

File: eth/watchers/orchestratorwatcher.go (hunk 4 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in 6a397072

File: eth/watchers/types.go (hunk 3 above)
Comment (kyriediculous) [reply eth/watchers/types.go]: fixed in 6a397072

File: eth/watchers/orchestratorwatcher.go (hunk 2 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in 6a397072

File: eth/watchers/orchestratorwatcher.go (hunk 1 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in 6a397072

**Comment by kyriediculous** [comment]
2ebd7fc8 fixes TranscoderDeactivated test for removed logs (stubClient.Orch was missing an Address field) and use `NewDBOrch()` instead of literal.

**Comment by kyriediculous** [comment]
9d860c1 properly sets an O's `deactivationRound` to the maximum possible value for an `int64` upon a `TranscoderActivated`  event (similar to how we set it to `2^256-1` in `BondingManager`), when a `TranscoderActivated` event is removed now also setting the `deactivationRound` retrieved from the RPC call on O.

Otherwise `deactivationRound` would be 0, but `currentRound < deactivationRound` would always return false, and thus an O would always be inactive.

File: eth/watchers/orchestratorwatcher.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+func (ow *OrchestratorWatcher) handleLog(log types.Log) error {
+	eventName, err := ow.dec.FindEventName(log)
+	if err != nil {
+		// Noop if we cannot find the event name
+		return nil
+	}
+
+	switch eventName {
+	case "TranscoderActivated":
+		if err := ow.handleTranscoderActivated(log); err != nil {
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Could shorten this to just `return ow.handleTranscoderActivated(log)`

File: eth/watchers/orchestratorwatcher.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+			return err
+		}
+	case "TranscoderDeactivated":
+		if err := ow.handleTranscoderDeactivated(log); err != nil {
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Could shorten this to just `return ow.handleTranscoderDeactivated(log)`

File: eth/watchers/orchestratorwatcher.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+		if err := ow.store.UpdateOrch(common.NewDBOrch("", transcoderDeactivated.Transcoder.String(), 0, 0, transcoderDeactivated.DeactivationRound.Int64())); err != nil {
+			return err
+		}
+	} else {
+		t, err := ow.lpEth.GetTranscoder(transcoderDeactivated.Transcoder)
+		if err != nil {
+			return err
+		}
+
+		if err := ow.store.UpdateOrch(common.NewDBOrch("", t.Address.String(), 0, 0, t.DeactivationRound.Int64())); err != nil {
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Looks like we set both the `activationRound` and `deactivationRound` when there is a removed `TranscoderActivated` log, but not here. Let's just be consistent and always set both the `activationRound` and `deactivationRound` if a log is removed?

File: eth/watchers/orchestratorwatcher.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+func (ow *OrchestratorWatcher) handleTranscoderActivated(log types.Log) error {
+	var transcoderActivated contracts.BondingManagerTranscoderActivated
+	if err := ow.dec.Decode("TranscoderActivated", log, &transcoderActivated); err != nil {
+		return err
+	}
+
+	if !log.Removed {
+		if err := ow.store.UpdateOrch(common.NewDBOrch("", transcoderActivated.Transcoder.String(), 0, transcoderActivated.ActivationRound.Int64(), maxFutureRound)); err != nil {
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Could remove the `err != nil` check here and just return

File: eth/watchers/orchestratorwatcher.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+			return err
+		}
+	} else {
+		t, err := ow.lpEth.GetTranscoder(transcoderActivated.Transcoder)
+		if err != nil {
+			return err
+		}
+		if err := ow.store.UpdateOrch(common.NewDBOrch("", t.Address.String(), 0, t.ActivationRound.Int64(), t.DeactivationRound.Int64())); err != nil {
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Could remove the `err != nil` check here and just return

File: eth/watchers/orchestratorwatcher.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+func (ow *OrchestratorWatcher) handleTranscoderDeactivated(log types.Log) error {
+	var transcoderDeactivated contracts.BondingManagerTranscoderDeactivated
+	if err := ow.dec.Decode("TranscoderDeactivated", log, &transcoderDeactivated); err != nil {
+		return err
+	}
+
+	if !log.Removed {
+		if err := ow.store.UpdateOrch(common.NewDBOrch("", transcoderDeactivated.Transcoder.String(), 0, 0, transcoderDeactivated.DeactivationRound.Int64())); err != nil {
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Could remove the `err != nil` check here and just return

File: eth/watchers/orchestratorwatcher.go (hunk 9 above)
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: Could remove the `err != nil` check here and just return

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: eth/watchers/orchestratorwatcher.go (hunk 9 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in d3fe26f

File: eth/watchers/orchestratorwatcher.go (hunk 12 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in d3fe26f

File: eth/watchers/orchestratorwatcher.go (hunk 11 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in d3fe26f

File: eth/watchers/orchestratorwatcher.go (hunk 10 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in d3fe26f

File: eth/watchers/orchestratorwatcher.go (hunk 9 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in d3fe26f

File: eth/watchers/orchestratorwatcher.go (hunk 8 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in d3fe26f

File: eth/watchers/orchestratorwatcher.go (hunk 7 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in d3fe26f

File: eth/watchers/orchestratorwatcher.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/golang/glog"
+	"github.com/livepeer/go-livepeer/common"
+	"github.com/livepeer/go-livepeer/eth"
+	"github.com/livepeer/go-livepeer/eth/blockwatch"
+	"github.com/livepeer/go-livepeer/eth/contracts"
+)
+
+const maxFutureRound = int64(1<<63 - 1)
```
Comment (yondonfu) [inline eth/watchers/orchestratorwatcher.go]: We can use `int64(math.MaxInt64)` here.

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: eth/watchers/orchestratorwatcher.go (hunk 13 above)
Comment (kyriediculous) [reply eth/watchers/orchestratorwatcher.go]: fixed in 0fa10821

**Review by yondonfu — COMMENTED** [review]
Changes look good - let's rebase

**Comment by kyriediculous** [comment]
rebased !

**Comment by yondonfu** [comment]
dc7b318 looks like a duplicate commit (already in #1171). Can we drop it?

**Comment by kyriediculous** [comment]
done !

**Review by yondonfu — APPROVED** [review]
LGTM 🚢
